### DISCUSSION TITLE
ssh 8.4p1 seems to be built wrong

### DISCUSSION THREAD

**Discussion by cellarweasel** [Homebrew/discussions#78] [Everyday usage] [answered]
Hello,
Tl;dr are other people having this issue? Or is it a weird my machine only thing: 
-------
after installing ssh through brew it fails, (I upgraded and updated twice like the instructions say to) 
```
╰─$ ssh                                                                                                                                               (base)
dyld: Library not loaded: /usr/local/opt/ldns/lib/libldns.3.dylib
  Referenced from: /usr/local/bin/ssh
  Reason: image not found
fish: 'ssh' terminated by signal SIGABRT (Abort)
```
`brew audit --strict openssh` reports the same issue. 
```
╰─$ brew audit --strict openssh                                                                                                                       (base)
Broken dependencies:
  /usr/local/opt/ldns/lib/libldns.3.dylib (ldns)
No broken library linkage detected
openssh:
  * openssh has broken dynamic library links:

    Rebuild this from source with:
      brew reinstall --build-from-source openssh
    If that's successful, file an issue here:
      https://github.com/Homebrew/homebrew-core/issues
Error: 1 problem in 1 formula detected
```
---
Doing some basic sleuthing reveals that ldns is essentially where openssh thinks it will be but libldns.0.dylib is installed as a dependency of openssh, and it doesn't suffice. 
Hacking it by symlinking lbldns.0.dylib to libdns.3.dylib makes openssh produce a very straight forward error: 
```
dyld: Library not loaded: /usr/local/opt/ldns/lib/libldns.3.dylib
  Referenced from: /usr/local/bin/ssh
  Reason: Incompatible library version: ssh requires version 4.0.0 or later, but libldns.3.dylib provides version 1.0.0
fish: 'ssh' terminated by signal SIGABRT (Abort)
```

**Comment by gromgit** [comment]
What does `brew info ldns` output?

**Reply by cellarweasel** [reply]
```
╭─edroot@s1016982 /usr/local/Cellar/ldns/1.7.1_2/lib
╰─$  brew info ldns                                                                                                                                   (base)
ldns: stable 1.7.1 (bottled)
DNS library written in C
https://nlnetlabs.nl/projects/ldns/
/usr/local/Cellar/ldns/1.7.1_2 (572 files, 5.7MB) *
  Poured from bottle on 2020-06-16 at 19:34:26
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/ldns.rb
==> Dependencies
Build: swig ✘
Required: openssl@1.1 ✔, python@3.8 ✔
==> Analytics
install: 3,587 (30 days), 10,893 (90 days), 56,435 (365 days)
install-on-request: 706 (30 days), 2,249 (90 days), 5,576 (365 days)
build-error: 0 (30 days)
```

Also I took the audit's advice and compiled it with `brew reinstall --build-from-source openssh`
and now everything works. It seems it was compiled on a machine that had libldns.3.dylib and there is some kind of #ifdef in the openssh code for ldns at version 4. 

I think this is a compile time issue not a formula issue.

**Comment by gromgit** [comment, marked as answer]
Given all you've mentioned, I suspect you have an old `ldns` bottle install for some reason. `/usr/local/opt/ldns/lib/libldns.3.dylib` should definitely exist in your current install, and it should definitely be version 4.0.0:
```
$ otool -L /usr/local/opt/ldns/lib/libldns.3.dylib
/usr/local/opt/ldns/lib/libldns.3.dylib:
	/usr/local/opt/ldns/lib/libldns.3.dylib (compatibility version 4.0.0, current version 4.0.0)
	/usr/local/opt/openssl@1.1/lib/libcrypto.1.1.dylib (compatibility version 1.1.0, current version 1.1.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.0.0)
```
You might want to force a reinstall of a fresh `ldns` bottle, then run the `otool` command above to verify correct installation:
```
brew fetch --force --force-bottle ldns
brew reinstall --force-bottle ldns
```

**Reply by cellarweasel** [reply]
Yeah ok! that fixes it also. Weird how I also got it working just fine by recompiling openssh with libldns.0.dylib. 

Ok well I don't know exactly how I got that half baked version of ldns, but I did have a `brew upgrade` go sideways when gcc took too long to compile and I cancelled it and uninstalled gcc. 
I wonder if I messed up brews internal consistency? :/

**Reply by SMillerDev** [reply]
Yeah, cancelling halfway through installs is generally a bad idea. Why are you compiling everything instead of using bottles?

**Reply by JJJ** [reply]
I just ran into this with a `brew upgrade` that upgraded `openssh` to "openssh-8.4p1_2.big_sur.bottle.tar.gz".

I also reinstalled `swig` to make the "✘" go away:

```
brew fetch --force --force-bottle swig
brew reinstall --force-bottle swig
```
